### PR TITLE
feat(melange): add `runtime_deps` to copy assets to the target dir

### PR DESCRIPTION
**Pull request by anmonteiro**
- as the test indicates, this isn't yet ready for consumption because dune currently lacks the ability to generate aliases for generated directories

### CODE REVIEW

File: src/dune_rules/dep_conf_eval.ml [hunk 1]
```diff
@@ -275,8 +275,11 @@ let named ~expander l =
...
-        let+ () = acc
-        and+ _x = to_action_builder (dep expander x) in
-        ())
+  ( (let+ paths =
+       List.fold_left l ~init:(Action_builder.return []) ~f:(fun acc x ->
+           let+ acc = acc
+           and+ paths = to_action_builder (dep expander x) in
+           paths :: acc)
+     in
+     List.concat paths)
```
Comment (rgrinberg) [inline src/dune_rules/dep_conf_eval.ml:284]: Doesn't seem good to collect all these paths everywhere if they're only used for melange. What about adding a separate function?

File: src/dune_rules/melange/melange_rules.ml [hunk 2]
```diff
@@ -241,6 +241,64 @@ let setup_emit_cmj_rules ~sctx ~dir ~scope ~expander ~dir_contents
...
+  let* melange_files =
+    let* expander = Super_context.expander sctx ~dir in
+    Runtime_deps.eval ~expander mel.runtime_deps
+  in
+  let targets = Runtime_deps.targets ~output melange_files in
+  let+ () =
+    let loc = mel.loc in
+    Memo.parallel_iter targets ~f:(fun (src, dst) ->
+        Super_context.add_rule ~loc ~dir ~mode sctx
+          (Action_builder.copy ~src ~dst))
```
Comment (rgrinberg) [inline src/dune_rules/melange/melange_rules.ml:288]: What about symlinking?

File: test/blackbox-tests/test-cases/melange/emit-with-runtime-deps.t [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  $ cat > dune-project <<EOF
+  > (lang dune 3.7)
+  > (using melange 0.1)
+  > EOF
+
+  $ cat > dune <<EOF
+  > (melange.emit
+  >  (alias mel)
+  >  (target output)
+  >  (runtime_deps assets/file.txt)
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/melange/emit-with-runtime-deps.t:12]: Another thing to test would be including directory targets in this list.

Comment (rgrinberg) [reply test/blackbox-tests/test-cases/melange/emit-with-runtime-deps.t:12]: One more test case: including the same path twice.

File: src/dune_rules/melange/melange_rules.ml [hunk 4]
```diff
@@ -241,6 +241,64 @@ let setup_emit_cmj_rules ~sctx ~dir ~scope ~expander ~dir_contents
...
+            lib_output_path ~output_dir ~lib_dir src
+        in
+        (src, dst))
+      deps
+end
+
+let setup_runtime_assets_rules sctx ~dir ~target_dir ~mode
+    ~(mel : Melange_stanzas.Emit.t) ~output =
+  let open Memo.O in
+  let* melange_files =
```
Comment (jchavarri) [inline src/dune_rules/melange/melange_rules.ml:279]: This naming got me confusing, the value contains a list of paths for the runtime deps right? What about `deps_paths`?

File: src/dune_rules/melange/melange_rules.ml [hunk 5]
```diff
@@ -241,6 +241,64 @@ let setup_emit_cmj_rules ~sctx ~dir ~scope ~expander ~dir_contents
...
+    let* expander = Super_context.expander sctx ~dir in
+    Runtime_deps.eval ~expander mel.runtime_deps
```
Comment (jchavarri) [inline src/dune_rules/melange/melange_rules.ml:281]: I understand this kind of expansion doesn't support globs right?

File: src/dune_rules/melange/melange_rules.ml [hunk 6]
```diff
@@ -275,13 +333,19 @@ let setup_entries_js ~sctx ~dir ~dir_contents ~scope ~compile_info ~target_dir
   in
   let output = `Private_library_or_emit target_dir in
   let obj_dir = Obj_dir.of_local obj_dir in
+  let* (_runtime_deps : unit Action_builder.t) =
```
Comment (jchavarri) [inline src/dune_rules/melange/melange_rules.ml:336]: I don't understand why `setup_runtime_assets_rules` returns something different than `unit` if it is not used.

File: test/blackbox-tests/test-cases/melange/emit-with-runtime-deps.t (hunk 3 above)
Comment (jchavarri) [reply test/blackbox-tests/test-cases/melange/emit-with-runtime-deps.t:12]: Maybe another test could be trying to "escape" the workspace with relative paths (not sure if this is covered already by some of the resolution functions being used)

**Review by jchavarri — APPROVED** [review]
Looks good. Besides the inlined questions, I wonder about some feature that is possible with current approach (`copy_files` + `deps`).

Inside `deps` field, one can leverage `glob_files_rec`, which is really useful when you have a lot of nested folders with multiple assets, e.g. `(glob_files_rec *.{svg,png})`.

Would something like that be possible with `runtime_deps`?

File: src/dune_rules/melange/melange_rules.ml (hunk 5 above)
Comment (anmonteiro) [reply src/dune_rules/melange/melange_rules.ml:281]: I'm pretty sure globs are supported. What made you think that?

File: src/dune_rules/melange/melange_rules.ml (hunk 6 above)
Comment (rgrinberg) [reply src/dune_rules/melange/melange_rules.ml:336]: They should just be attached to the alias. Otherwise, it would slow down the build.

File: src/dune_rules/melange/melange_rules.ml (hunk 2 above)
Comment (anmonteiro) [reply src/dune_rules/melange/melange_rules.ml]: great idea, done in [bef37ff](https://github.com/ocaml/dune/pull/7199/commits/bef37ffb47d1133fb3864e7778726bca8d88fe41)

File: src/dune_rules/dep_conf_eval.ml (hunk 1 above)
Comment (anmonteiro) [reply src/dune_rules/dep_conf_eval.ml]: added in [e679967](https://github.com/ocaml/dune/pull/7199/commits/e679967535f2f426aa268a971c6b99674ce4cb6a)

File: test/blackbox-tests/test-cases/melange/emit-with-runtime-deps.t (hunk 3 above)
Comment (anmonteiro) [reply test/blackbox-tests/test-cases/melange/emit-with-runtime-deps.t:12]: Added edge case tests here: [7ab3576](https://github.com/ocaml/dune/pull/7199/commits/7ab3576e0579e38e3389d1d6dec548d0fdee7059)

Added a directory target test showing it doesn't work: [99c648c](https://github.com/ocaml/dune/pull/7199/commits/99c648c6758897be4fe4a65e48f9016a0249143f)

File: test/blackbox-tests/test-cases/melange/emit-with-runtime-deps-dir-target.t [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+          melc output/main.js
+  File "dune", line 5, characters 0-97:
+  5 | (melange.emit
+  6 |  (alias mel)
+  7 |  (target output)
+  8 |  (runtime_deps ./some_dir)
+  9 |  (module_system commonjs))
+  Error: Error trying to read targets after a rule was run:
+  - output/some_dir: Unexpected file kind "S_DIR" (directory)
+  [1]
```
Comment (anmonteiro) [inline test/blackbox-tests/test-cases/melange/emit-with-runtime-deps-dir-target.t:45]: With copy instead of symlink, the error is:

```
+  Error: Is a directory
+  -> required by _build/default/output/some_dir
+  -> required by alias output/mel
```

File: src/dune_rules/melange/melange_rules.ml (hunk 4 above)
Comment (anmonteiro) [reply src/dune_rules/melange/melange_rules.ml]: good point, renamed in [b1df820](https://github.com/ocaml/dune/pull/7199/commits/b1df8200b43295b8f521e8f0a1b00ba3f7425f95)

**Comment by anmonteiro** [comment]
> Would something like that be possible with runtime_deps?

Globbing should work, I'll make it clear in a test.

File: src/dune_rules/melange/melange_rules.ml (hunk 5 above)
Comment (anmonteiro) [reply src/dune_rules/melange/melange_rules.ml:283]: check out [01429cf](https://github.com/ocaml/dune/pull/7199/commits/01429cfb277cd04b70b40b861717b67b37fcbe5e) for an example of globs being supported.

File: test/blackbox-tests/test-cases/melange/emit-with-runtime-deps-dir-target.t (hunk 7 above)
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/melange/emit-with-runtime-deps-dir-target.t:45]: Interesting. So this is only a problem with symlinking? I would expect the same problem with copying. For symlinking, we have `Action_builder.symlink_dir` to get around this.

Comment (anmonteiro) [reply test/blackbox-tests/test-cases/melange/emit-with-runtime-deps-dir-target.t:45]: It looks like a problem with copying too, right? where copying doesn't pass `-r`

File: src/dune_rules/melange/melange_rules.ml (hunk 5 above)
Comment (jchavarri) [reply src/dune_rules/melange/melange_rules.ml:283]: > I'm pretty sure globs are supported. What made you think that?

In [my attempt to implement runtime_deps](https://github.com/ocaml/dune/pull/6643), I took `Simple_rules.copy_files` as reference on how to handle globs. As one can see, that function handles a lot of edge cases like "Cannot find directory:" or "Cannot copy files onto themselves...". As I didn't see any of this handling in the PR, I assumed globs were not supported. Isn't it necessary to handle these cases here?

File: test/blackbox-tests/test-cases/melange/emit-with-runtime-deps-dir-target.t (hunk 7 above)
Comment (anmonteiro) [reply test/blackbox-tests/test-cases/melange/emit-with-runtime-deps-dir-target.t:45]: my code is probably missing a directory check, and call `with_targets ~directory_targets...`  rather than the current form. 

Which brings up the question: what's the best way to detect whether I'm looking at a directory or a file in an action builder monad?

File: src/dune_rules/melange/melange_rules.ml (hunk 5 above)
Comment (anmonteiro) [reply src/dune_rules/melange/melange_rules.ml:283]: oh interesting. I might need to do all of that indeed.

File: test/blackbox-tests/test-cases/melange/emit-with-runtime-deps-dir-target.t (hunk 7 above)
Comment (jchavarri) [reply test/blackbox-tests/test-cases/melange/emit-with-runtime-deps-dir-target.t:45]: > For symlinking, we have `Action_builder.symlink_dir` to get around this.

What would happen if user tries to copy a dir and also another dir inside it?

Comment (anmonteiro) [reply test/blackbox-tests/test-cases/melange/emit-with-runtime-deps-dir-target.t:45]: I believe directory targets are opaque, so that works.

Comment (rgrinberg) [reply test/blackbox-tests/test-cases/melange/emit-with-runtime-deps-dir-target.t:45]: Not sure if there's a good way to check . For now, I'd do the following:

* Use symlinking
* Document that directories are a known limitation

**Comment by rgrinberg** [comment]
could you rebase this?

File: src/dune_rules/dep_conf_eval.ml [hunk 8]
```diff
@@ -273,10 +273,19 @@ let named ~expander l =
...
+           and+ paths = to_action_builder (dep expander x) in
+           paths :: acc)
+     in
+     Path.Set.of_list (List.concat paths))
   , List.fold_left l ~init:sandbox ~f:add_sandbox_config )
+
+let unnamed ?(sandbox = Sandbox_config.no_special_requirements) ~expander l =
+  let expander = prepare_expander expander in
+  let builder, sandbox = unnamed_get_paths ~sandbox ~expander l in
+  (Action_builder.ignore builder, sandbox)
```
Comment (rgrinberg) [inline src/dune_rules/dep_conf_eval.ml:291]: Although it requires a bit of duplication, I'd prefer to keep the old implementation of unnamed. This new implementation has the potential to really slow down old builds with rules that expand huge globs.

File: src/dune_rules/foreign_rules.ml [hunk 9]
```diff
@@ -252,7 +252,11 @@ let build_o_files ~sctx ~foreign_sources ~(dir : Path.Build.t) ~expander
       in
       let extra_deps, sandbox =
         match src.kind with
-        | Stubs stubs -> Dep_conf_eval.unnamed stubs.extra_deps ~expander
+        | Stubs stubs ->
+          let action_builder, sandbox =
+            Dep_conf_eval.unnamed stubs.extra_deps ~expander
+          in
+          (Action_builder.ignore action_builder, sandbox)
```
Comment (rgrinberg) [inline src/dune_rules/foreign_rules.ml:259]: I think this can be reverted now.

File: src/dune_rules/mdx.ml [hunk 10]
```diff
@@ -345,7 +345,7 @@ let gen_rules_for_single_file stanza ~sctx ~dir ~expander ~mdx_prog
       let deps, sandbox =
         Dep_conf_eval.unnamed ~expander (mdx_package_deps @ mdx_generic_deps)
       in
-      Action_builder.with_no_targets deps
+      Action_builder.with_no_targets (Action_builder.ignore deps)
```
Comment (rgrinberg) [inline src/dune_rules/mdx.ml:348]: Same as this

File: src/dune_rules/melange/melange_rules.ml [hunk 11]
```diff
@@ -241,6 +241,64 @@ let setup_emit_cmj_rules ~sctx ~dir ~scope ~expander ~dir_contents
   in
   Buildable_rules.with_lib_deps ctx compile_info ~dir ~f
 
+module Runtime_deps = struct
+  let to_action_builder ~expander dep_conf =
+    let runtime_deps =
+      let runtime_deps, _sandbox = Dep_conf_eval.unnamed ~expander dep_conf in
+      runtime_deps
+    in
+    Action_builder.memoize "melange runtime deps" runtime_deps
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/dune_rules/melange/melange_rules.ml:258]: Returning Memo.t doesn't mean the returned value is memoized. It just means that it might possibly be memoized. I agree though, I'm not sure I see the need to memoize the set of deps here. I don't see it being reused anywhere.

File: src/dune_rules/melange/melange_rules.ml [hunk 12]
```diff
@@ -247,6 +247,66 @@ let setup_emit_cmj_rules ~sctx ~dir ~scope ~expander ~dir_contents
...
+            let target =
+              src |> Path.as_in_build_dir_exn
+              |> Path.Build.drop_build_context_exn
+            in
+            Path.Build.append_source output_dir target
+          | `Public_library (lib_dir, output_dir) ->
+            lib_output_path ~output_dir ~lib_dir src
+        in
+        (src, dst))
+      deps
```
Comment (rgrinberg) [inline src/dune_rules/melange/melange_rules.ml:281]: You can put `deps` before `~f` and shave off a line of code.

File: src/dune_rules/melange/melange_rules.ml [hunk 13]
```diff
@@ -247,6 +247,66 @@ let setup_emit_cmj_rules ~sctx ~dir ~scope ~expander ~dir_contents
...
+    paths
+
+  let targets ~output deps =
+    Path.Set.to_list_map
+      ~f:(fun src ->
+        let dst =
+          match output with
+          | `Private_library_or_emit output_dir ->
+            let target =
+              src |> Path.as_in_build_dir_exn
```
Comment (rgrinberg) [inline src/dune_rules/melange/melange_rules.ml:273]: I believe this would fail if there are any external paths in runtime_deps. Worth adding a test for this.

**Comment by anmonteiro** [comment]
Thanks for your comments -- addressed them and this should be ready for another look

File: src/dune_rules/melange/melange_rules.ml [hunk 14]
```diff
@@ -247,6 +247,59 @@ let setup_emit_cmj_rules ~sctx ~dir ~scope ~expander ~dir_contents
...
+  let targets ~output deps =
+    Path.Set.to_list deps
+    |> List.filter_map ~f:(fun src ->
+           match output with
+           | `Private_library_or_emit output_dir -> (
+             match Path.as_in_build_dir src with
+             | Some src_build ->
+               let target = Path.Build.drop_build_context_exn src_build in
+               Some (src, Path.Build.append_source output_dir target)
+             | None -> None)
```
Comment (rgrinberg) [inline src/dune_rules/melange/melange_rules.ml:272]: Instead of dropping, you should add them to as deps to the alias directly without the copying rules

Comment (anmonteiro) [reply src/dune_rules/melange/melange_rules.ml:272]: ah that makes sense

Comment (anmonteiro) [reply src/dune_rules/melange/melange_rules.ml]: fixed this one in [76ef993](https://github.com/ocaml/dune/pull/7199/commits/76ef993a276e3abb54bf1a22408ff113f34bb6cd)
